extensions/native/circuit/src/sumcheck/air.rs

```diff
@@ -0,0 +1,584 @@
+        //         -> logup_row -> ... -> logup_row
+        ////////////////////////////////////////////////////////////////
+
+        // (curr_prod_n, curr_logup_n) start at 0
+        builder.when(header_row).assert_zero(curr_prod_n);
+        builder
+            .when(header_row + prod_row)
+            .assert_zero(curr_logup_n);
+        builder
+            .when(next.prod_row)
+            .assert_eq(curr_prod_n + AB::F::ONE, next.curr_prod_n);
+        builder
+            .when(next.logup_row)
+            .assert_eq(curr_logup_n + AB::F::ONE, next.curr_logup_n);
+        // if header row is followed by another header row
+        // then num_prod_spec and num_logup_spec should be zero
+        builder
+            .when(header_row)
+            .when(next.header_row)
+            .assert_zero(num_prod_spec);
+        builder
+            .when(header_row)
+            .when(next.header_row)
+            .assert_zero(num_logup_spec);
+        // if header row is followed by a logup row,
+        // then num_prod_spec should be zero
+        builder
+            .when(header_row)
+            .when(next.logup_row)
+            .assert_zero(num_prod_spec);
+        builder
+            .when(prod_row)
+            .when(next.logup_row)
+            .assert_eq(curr_prod_n, num_prod_spec);
+        builder
+            .when(logup_row)
+            .when(next.header_row)
+            .assert_eq(curr_logup_n, num_logup_spec);
+
+        // Timestamp transition
+        builder
+            .when(header_row)
+            .when(next.prod_row + next.logup_row)
+            .assert_eq(
+                next.start_timestamp,
+                start_timestamp + AB::F::from_canonical_usize(7),
+            );
+        builder
+            .when(prod_row)
+            .when(next.prod_row + next.logup_row)
+            .assert_eq(
+                next.start_timestamp,
+                start_timestamp + AB::F::ONE + within_round_limit * AB::F::TWO,
+            );
+        builder
+            .when(logup_row)
+            .when(next.prod_row + next.logup_row)
+            .assert_eq(
+                next.start_timestamp,
+                start_timestamp + AB::F::ONE + within_round_limit * AB::F::from_canonical_usize(3),
+            );
+
+        // Termination condition
+        // TODO: enable this
+        // assert_array_eq(
+        //     &mut builder.when::<AB::Expr>(is_end.into()),
+        //     eval_acc,
+        //     [AB::F::ZERO; 4],
+        // );
+
+        // Randomness transition
+        assert_array_eq(
+            &mut builder.when(and(header_row, next.prod_row + next.logup_row)),
+            next.challenges[0..EXT_DEG].try_into().unwrap(),
+            [AB::F::ONE, AB::F::ZERO, AB::F::ZERO, AB::F::ZERO],
+        );
+        assert_array_eq::<_, _, _, { EXT_DEG }>(&mut builder.when(header_row), alpha, alpha1);
+        let prod_next_alpha = FieldExtension::multiply(alpha1, alpha);
+        // TODO: reduce the degree
```

Comment:
TODO: add one column to store `prod_row * next.prod_row` as this condition's degree is two.